// audio-effects/dist/index.js
"use strict";
/**
 * Export all effects:
 * - Input
 * - Output
 * - Volume
 * - Distortion
 * - Delay
 * - Flanger
 * - Reverb
 * - Tremolo
 */
var Input_1 = require('./audio-nodes/effects/Input');
exports.Input = Input_1.Input;
var Output_1 = require('./audio-nodes/effects/Output');
exports.Output = Output_1.Output;
var Volume_1 = require('./audio-nodes/effects/Volume');
exports.Volume = Volume_1.Volume;
var Distortion_1 = require('./audio-nodes/effects/Distortion');
exports.Distortion = Distortion_1.Distortion;
var Delay_1 = require('./audio-nodes/effects/Delay');
exports.Delay = Delay_1.Delay;
var Flanger_1 = require('./audio-nodes/effects/Flanger');
exports.Flanger = Flanger_1.Flanger;
var Reverb_1 = require('./audio-nodes/effects/Reverb');
exports.Reverb = Reverb_1.Reverb;
var Tremolo_1 = require('./audio-nodes/effects/Tremolo');
exports.Tremolo = Tremolo_1.Tremolo;
/**
 * Export the base audioNodes:
 * - SingleAudioNode
 * - MultiAudioNode
 */
var SingleAudioNode_1 = require('./audio-nodes/SingleAudioNode');
exports.SingleAudioNode = SingleAudioNode_1.SingleAudioNode;
var MultiAudioNode_1 = require('./audio-nodes/MultiAudioNode');
exports.MultiAudioNode = MultiAudioNode_1.MultiAudioNode;
/**
 * Export helper-functions:
 *  - HasAudioContext: Check if the current browser supports the web-audio-api.
 *  - HasGetUserMedia: Check if the current browser supports getUserMedia.
 */
var hasAudioContext_1 = require('./helpers/hasAudioContext');
exports.HasAudioContext = hasAudioContext_1.HasAudioContext;
var hasGetUserMedia_1 = require('./helpers/hasGetUserMedia');
exports.HasGetUserMedia = hasGetUserMedia_1.HasGetUserMedia;
